src/bsoncxx/builder/list.hpp

```diff
@@ -181,7 +153,32 @@ class array : public list {
     /// @see bsoncxx::builder::list
     /// @see bsoncxx::builder::document
     ///
-    array(initializer_list_t init) : list(init, false, true) {}
+    array(initializer_list_t init) : _is_array{true} {
+        for (auto&& ele : init)
+            append_array_or_value(_core, ele);
+    }
+
+    operator bsoncxx::array::value() {
```

Comment:
As nice as this is, I think we should require users to explicitly `extract()` their final documents.  There are historically a lot of instances of the C++ driver allowing people to shoot themselves in the foot with owned vs. non-owned things, so I think it's best to be really explicit about it.  Even though this doesn't return something view-y and sneaky, I would vote for mandating a call to `extract()` rather than providing implicit conversion.  We could always add this as a nice-to-have in the future, if it feels right.

Reply:
If this were explicit, then this
```
// { "0" : "foo", "1" : [ "bar", "baz" ], "2" : "qux" }
array arr = {"foo", array{"bar", "baz"}, "qux"};
```
would have to become this
```
array nested = {"bar", "baz"}
array arr = {"foo", nested.extract(), "qux"};
```
The former only works because every element can be used to direct-initialize a bson_value::value. If this were explicit, this would no longer be the case. Is there a specific example you could see this causing an issue for?

Reply:
Edit: done.